src/nmodl/nocpout.cpp

```diff
@@ -3135,3 +3135,116 @@ Symbol* breakpoint_current(Symbol* s) {
     }
     return s;
 }
+
+// Determine if setdata is required to call FUNCTION or PROCEDURE
+// setdata is required if RANGE var used. For safety, also VERBATIM.
+// Deal with nested calls, via maintaining a list for each func.
+// Note that the nest can be recursive and called function may not
+// yet be defined til entire text is processed.
+
+#include <unordered_set>
+#include <unordered_map>
+
+typedef struct {
+    std::unordered_set<Symbol*> func_calls;
+    bool need_setdata{false};
+    bool is_being_looked_at{false};  // avoid recursion loops
+    Item* q{nullptr};                // To be modified if need_setdata.
+} Info;
+
+static std::unordered_map<Symbol*, Info> funcs;
+static Symbol* in_func_;
+
+void check_range_in_func(Symbol* s) {
+    if (in_func_) {
+        // If s is a RANGE variable or nullptr (VERBATIM)
+        // then mark the current function as needing setdata
+        // If s is FUNCTION or PROCEDURE, then add to list
+        Info& i = funcs[in_func_];
+        if (!s) {  // VERBATIM
+            i.need_setdata = true;
+        } else if (s->nrntype & (NRNRANGE | NRNPOINTER)) {
+            i.need_setdata = true;
+        } else if (s->usage & FUNCT) {
+            i.func_calls.insert(s);
+        }
+    }
+}
+
+void set_inside_func(Symbol* s) {
+    in_func_ = s;
+    if (s) {
+        assert(funcs.count(s) == 0);
+        Info i{};
+        funcs[s] = i;
+    }
+}
+
+// Make sure need_setdata is properly marked for all funcs.
+// I.e on entry, only ones marked are those that use RANGE or VERBATIM.
+// Need to recursively look through func_calls but watch out for loops.
+
+static bool check_func(Symbol* s);  // recursive
+
+void func_needs_setdata() {
+    if (strcmp(mechname, "nothing") == 0) {
+        return;
+    }
+    // No _extcall_prop for non vectorized density mechanisms.
+    if (!vectorize) {
+        return;
+    }
+    for (auto& f: funcs) {
+        f.second.is_being_looked_at = false;
+    }
+    for (auto& f: funcs) {
+        check_func(f.first);
+    }
+    for (auto& f: funcs) {  // update the hocfunc item if need_setdata
+        auto& q = f.second.q;
+        if (f.second.q && f.second.need_setdata) {
+            Symbol* s = f.first;
+            sprintf(buf,
+                    "\n  if(!_extcall_prop) { hoc_execerror(\""
```

Comment:
non-vectorized don't use _extcall_prop (which is initalized to nullptr) but _iml which is initialized to 0. But if there are no items, 0 will cause segfault. Hmm. Wonder if we should initialize to -1.
The serious issue with both left over ```_extcall_prop``` or ```_iml``` is that either could become invalid, e.g. all instances deleted.